packages/tpetra/core/src/Tpetra_Details_ExecutionSpaces.hpp

```diff
@@ -0,0 +1,476 @@
+#ifndef TPETRA_DETAILS_SPACES_HPP
+#define TPETRA_DETAILS_SPACES_HPP
+
+#include <iostream>
+#include <sstream>
+#include <vector>
+
+#include <Kokkos_Core.hpp>
+
+#include <Teuchos_RCP.hpp>
+
+#include "Tpetra_Details_Behavior.hpp"
+#include "Tpetra_Details_Profiling.hpp"
+
+/*! \file
+
+Interface for Tpetra's managed Kokkos execution spaces
+This facility strives to provide:
+1. Caching previously-constructed spaces so they can be used on-demand in parts
+of Tpetra
+2. Spaces of different priorities where supported
+3. Fast sync of spaces where supported
+
+For each Kokkos backend, there is a singleton instance manager.
+This singleton manages the lifetime of all instances for that Kokkos backend.
+These singletons are not meant to be accessed directly, but through
+top-level functions in this file.
+
+*/
+
+#define TPETRA_DETAILS_SPACES_THROW(x)                                         \
+  {                                                                            \
+    std::stringstream ss;                                                      \
+    ss << __FILE__ << ":" << __LINE__ << ": " << x;                            \
+    throw std::runtime_error(ss.str());                                        \
+  }
+
+namespace Tpetra {
+namespace Details {
+namespace Spaces {
+
+/*! \brief Priority interface for Tpetra's managed execution spaces
+
+    Priority is best-effort. low <= medium <= high, but it may be the case that
+   priority levels are equivalent.
+*/
+enum class Priority {
+  low = 0,
+  medium = 1,
+  high = 2,
+  NUM_LEVELS = 3 // not to be used as a priority
+};
+
+#if defined(KOKKOS_ENABLE_CUDA)
+inline void success_or_throw(cudaError_t err, const char *file,
+                             const int line) {
+  if (err != cudaSuccess) {
+    std::stringstream ss;
+    ss << file << ":" << line << ": ";
+    ss << cudaGetErrorString(err);
+    throw std::runtime_error(ss.str());
+  }
+}
+#define CUDA_RUNTIME(x)                                                        \
+  Tpetra::Details::Spaces::success_or_throw((x), __FILE__, __LINE__)
```

Comment:
How about `TPETRA_DETAILS_SPACES_CUDA_RUNTIME`? A bit clumsy, but the hope is this would eventually be a part of Kokkos and this code could be deleted.